### PR TITLE
IBX-1694: Implemented Symfony Service Tag compatibility layer

### PR DESCRIPTION
**Pull request by alongosz**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-1694](https://issues.ibexa.co/browse/IBX-1694)
| **Type**                                   | feature
| **Target Ibexa version** | `v4.0`
| **BC breaks**                          | no

This PR adds compatibility layer for the legacy eZ Platform Symfony Service Container Tag names, defining the mapping between old names and new Ibexa names.

The solution is to find all legacy tags and add another tag with the new name and the same attributes. Legacy tag names are not removed from the definitions due to BC. Even if we ourselves remove old tags, there's no way to know if some 3rd party still relies on them.

- [x] What needs to be discussed here are the new naming choices made here. Grouping is done by the dots (`.`) in the new tag names, but groups are related to domain/functionality rather than a package defining the tag. If there's a certain pattern associated with a tag (e.g.: visitor, handler), it is placed at the end of a name.

### Naming convention
```yaml
ibexa.<domain>.[adapter].<specifics>.[design_pattern]
```
where `<specifics>` may or may not be grouped (aiming not to group) per case basis. If a group can be identified for certain items, it should be used. `[]` implies optional part.

### Previously deprecated tags

The support for Tags that were deprecated in Ibexa 3.3 or earlier is going to be dropped. They're not included in the map. Some leftover usages will be replaced across the codebase with the new name.

- `ezpublish.searchEngineIndexer`
- `ezpublish.searchEngine`
- `ezpublish.fieldType.externalStorageHandler`
- `ezpublish.fieldType.externalStorageHandler.gateway`
- `ezpublish.fieldType`
- `ezpublish_rest.field_type_processor`
- `ezpublish.fieldType.parameterProvider`
- `ezpublish.storageEngine.legacy.converter`
- `ezpublish.fieldType.indexable`
- `ez.fieldFormMapper.definition`
- `ez.fieldFormMapper.value`

### TODO
- [x] Update mapping with all needed tags
- [x] Add minimal test coverage

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review

### CODE REVIEW

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    'ezplatform.admin_ui.component' => 'ibexa.admin_ui.component',
+    'ezplatform.admin_ui.config_provider' => 'ibexa.admin_ui.config.provider',
+    'ezplatform.admin_ui.form_ui_action_mapper' => 'ibexa.admin_ui.form_ui.mapper.action',
+    'ezplatform.admin_ui.siteaccess_preview_voter' => 'ibexa.admin_ui.site_access.preview.voter',
+    'ezplatform.admin_ui.user_setting.form_mapper' => 'ibexa.admin_ui.user_setting.mapper.form',
+    'ezplatform.admin_ui.user_setting.value' => 'ibexa.admin_ui.user_setting.value',
+    'ezplatform.cache_response_tagger' => 'ibexa.cache.http.response.tagger',
+    'ezplatform.calendar.event_source' => 'ibexa.calendar.event.source',
+    'ezplatform.calendar.event_type.future_hide.action' => 'ibexa.calendar.event.type.action.future_hide',
+    'ezplatform.calendar.event_type.future_publication.action' => 'ibexa.calendar.event_type.action.future_publication',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]:
```suggestion
    'ezplatform.calendar.event_type.future_publication.action' => 'ibexa.calendar.event.type.action.future_publication',
```

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    'ezplatform.calendar.event_type' => 'ibexa.calendar.event.type',
+    'ezplatform.cron.job' => 'ibexa.cron.job',
+    'ezplatform.ezrichtext.configuration.provider' => 'ibexa.field_type.richtext.configuration.provider',
+    'ezplatform.field_type.comparable.engine' => 'ibexa.field_type.comparable.engine',
+    'ezplatform.field_type.comparable' => 'ibexa.field_type.comparable',
+    'ezplatform.field_type.external_storage_handler.gateway' => 'ibexa.field_type.storage.external.handler.gateway',
+    'ezplatform.field_type.external_storage_handler' => 'ibexa.field_type.storage.external.handler',
+    'ezplatform.field_type.form_mapper.definition' => 'ibexa.admin_ui.field_type.form_mapper.definition',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]:
```suggestion
    'ezplatform.field_type.form_mapper.definition' => 'ibexa.admin_ui.field_type.form.mapper.definition',
```
Not sure what is the general convention here, but I assume we don't separate core mechanism names like `FieldType`. However, if `EventType` is tagged as `event.type`, I would follow that rule also for mappers.

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    'ezplatform.field_type.form_mapper.value' => 'ibexa.admin_ui.field_type.form_mapper.value',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]:
```suggestion
    'ezplatform.field_type.form_mapper.value' => 'ibexa.admin_ui.field_type.form.mapper.value',
```
Same as above.

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    'ezplatform.field_type' => 'ibexa.field_type',
+    'ezplatform.field_type.indexable' => 'ibexa.field_type.indexable',
+    'ezplatform.field_type.legacy_storage.converter' => 'ibexa.field_type.storage.legacy.converter',
+    'ezplatform.fieldtype.page.relation.attribute_type_extractor' => 'ibexa.field_type.page.relation.attribute_type_extractor',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]:
```suggestion
    'ezplatform.fieldtype.page.relation.attribute_type_extractor' => 'ibexa.field_type.page.relation.attribute.type.extractor',
```
The same concern applies here - I see you wanted to group tags by domain/functionality but not seeing an obvious rule here is a bit confusing for me and might be inconvenient for developers.

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    'ezplatform.fieldtype.page.schedule_block.event_processor' => 'ibexa.field_type.page.schedule_block.event_processor',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]:
```suggestion
    'ezplatform.fieldtype.page.schedule_block.event_processor' => 'ibexa.field_type.page.schedule_block.event.processor',
```

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    'ezplatform.fieldtype.page.serializer.handler' => 'ibexa.field_type.page.serializer.handler',
+    'ezplatform.field_type.parameter_provider' => 'ibexa.field_type.view.parameter.provider',
+    'ezplatform.filter.criterion.query_builder' => 'ibexa.filter.criterion.query_builder',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]: Same here.

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    'ezplatform.filter.sort_clause.query_builder' => 'ibexa.filter.sort_clause.query_builder',
+    'ezplatform.form_builder.attribute_form_type_mapper' => 'ibexa.form_builder.form.type.attribute.mapper',
+    'ezplatform.form_builder.field_mapper' => 'ibexa.form_builder.field.mapper',
+    'ezplatform.form_builder.field_submission_converter' => 'ibexa.form_builder.field.submission.converter',
+    'ezplatform.form_builder.validator_constraint_mapper' => 'ibexa.form_builder.field.constraint.validator.mapper',
+    'ezplatform.form_builder.validator_form_type_mapper' => 'ibexa.form_builder.form.type.validator.mapper',
+    'ezplatform.form.trash_location_option' => 'ibexa.form.trash.location.option',
+    'ezplatform_graphql.domain_schema_worker' => 'ibexa.graphql.domain_schema.worker',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]:
```suggestion
    'ezplatform_graphql.domain_schema_worker' => 'ibexa.graphql.domain.schema.worker',
```

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php (hunk 2 above)
Comment (alongosz) [reply src/bundle/Resources/mappings/symfony-service-tag-name-map.php]: > Not sure what is the general convention here, but I assume we don't separate core mechanism names like `FieldType`. However, if `EventType` is tagged as `event.type`, I would follow that rule also for mappers.

Makes sense :+1:

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php (hunk 4 above)
Comment (alongosz) [reply src/bundle/Resources/mappings/symfony-service-tag-name-map.php]: > The same concern applies here - I see you wanted to group tags by domain/functionality but not seeing an obvious rule here is a bit confusing for me and might be inconvenient for developers.

`extractor` definitely should be separated, nice catch. As for attribute type for me it kinda goes together, but if it's confusing then I'm gonna go with your suggestion.

**Comment by alongosz** [comment]
@konradoboza https://github.com/ibexa/compatibility-layer/pull/12#pullrequestreview-833713747 resolved via 914caf7. Thanks!

**Comment by konradoboza** [comment]
Yet to be discussed, but it feels like we should go either with `.` or `_` as a separator in general. Having one specific pattern would speed up the process as end developers won't be bothered with any conventions at all while tagging. Determining what should "go together" in naming is relative, it feels like we would benefit from a rule of thumb here. Despite documenting this change I think it will be easier to just assume, that one rule is reused across the system and it's safe to follow it whenever service tags are applied.

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    'ezplatform.persistence.site_factory.criterion_handler' => 'ibexa.site.factory.storage.legacy.criterion.handler',
+    'ezplatform.query_type' => 'ibexa.query_type',
+    'ezplatform.query_type.sort_clause_parser' => 'ibexa.query_type.sort_clause.parser',
+    'ezplatform.search.elasticsearch.query.content.aggregation_result_extractor' => 'ibexa.search.elasticsearch.query.content.aggregation.result_extractor',
+    'ezplatform.search.elasticsearch.query.content.aggregation_visitor' => 'ibexa.search.elasticsearch.query.content.aggregation.visitor',
+    'ezplatform.search.elasticsearch.query.content.criterion_visitor' => 'ibexa.search.elasticsearch.query.content.criterion.visitor',
+    'ezplatform.search.elasticsearch.query.content.facet_builder_visitor' => 'ibexa.search.elasticsearch.query.content.facet.builder.visitor',
+    'ezplatform.search.elasticsearch.query.content.sort_clause_visitor' => 'ibexa.search.elasticsearch.query.content.sort_clause.visitor',
+    'ezplatform.search.elasticsearch.query.facet_result_extractor' => 'ibexa.search.elasticsearch.query.facet.result_extractor',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]: Why not `result.extractor` given we follow this approach for e.g. `CriterionVisitor` interface?

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    'ezplatform.search.elasticsearch.query.location.aggregation_result_extractor' => 'ibexa.search.elasticsearch.query.location.aggregation.result_extractor',
+    'ezplatform.search.elasticsearch.query.location.aggregation_visitor' => 'ibexa.search.elasticsearch.query.location.aggregation.visitor',
+    'ezplatform.search.elasticsearch.query.location.criterion_visitor' => 'ibexa.search.elasticsearch.query.location.criterion.visitor',
+    'ezplatform.search.elasticsearch.query.location.facet_builder_visitor' => 'ibexa.search.elasticsearch.query.location.facet.builder.visitor',
+    'ezplatform.search.elasticsearch.query.location.sort_clause_visitor' => 'ibexa.search.elasticsearch.query.location.sort_clause_visitor',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]: Same question as above.

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    'ezpublish.core.io.migration.file_lister' => 'ibexa.io.migration.file_lister',
+    'ezpublish_data_collector' => 'ibexa.debug.data_collector',
+    'ezpublish.limitationType' => 'ibexa.permissions.limitation_type',
+    'ezpublish.notification.renderer' => 'ibexa.notification.renderer',
+    'ezpublish.persistence.legacy.role.limitation.handler' => 'ibexa.storage.legacy.role.limitation.handler',
+    'ezpublish.persistence.legacy.url.criterion_handler' => 'ibexa.storage.legacy.url.criterion.handler',
+    'ezpublish.placeholder_provider' => 'ibexa.media.images.placeholder.provider',
+    'ezpublish_rest.input.handler' => 'ibexa.rest.input.handler',
+    'ezpublish_rest.input.parser' => 'ibexa.rest.input.parser',
+    'ezpublish_rest.output.value_object_visitor' => 'ibexa.rest.output.value_object_visitor',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php]: Same here.

File: src/bundle/Resources/mappings/symfony-service-tag-name-map.php [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    'ezpublish.search.legacy.gateway.sort_clause_handler.location' => 'ibexa.search.legacy.gateway.sort_clause_handler.location',
+    'ezpublish.search.solr.core_filter' => 'ibexa.search.solr.core.filter',
+    'ezpublish.search.solr.endpoint' => 'ibexa.search.solr.endpoint',
+    'ezpublish.search.solr.field_mapper.block' => 'ibexa.search.solr.field.mapper.block',
+    'ezpublish.search.solr.field_mapper.block_translation' => 'ibexa.search.solr.field.mapper.block.translation',
+    'ezpublish.search.solr.field_mapper.content' => 'ibexa.search.solr.field.mapper.content',
+    'ezpublish.search.solr.field_mapper.content_translation' => 'ibexa.search.solr.field.mapper.content.translation',
+    'ezpublish.search.solr.field_mapper.location' => 'ibexa.search.solr.field.mapper.location',
+    'ezpublish.search.solr.gateway' => 'ibexa.search.solr.gateway',
+    'ezpublish.search.solr.query.content.criterion_visitor' => 'ibexa.search.solr.query.content.criterion.visitor',
```
Comment (konradoboza) [inline src/bundle/Resources/mappings/symfony-service-tag-name-map.php:116]: Here, on the other hand, we follow "dot notation".

**Review by konradoboza — COMMENTED** [review]
The convention is still a little bit blurry to me, but the general naming looks fine. Few concerns:

**Review by konradoboza — APPROVED** [review]

**Comment by alongosz** [comment]
Update: fc42c63: removed `ezsystems.platformui.application_config_provider` from the map as it seems to be an unused leftover from v1.x. There are tagged services (ibexa/compatibility-layer#12), but tag itself is not processed anywhere.

**Review by ViniTou — APPROVED** [review]

**Pull request merged**